dynamic_programming/levenshtein_distance.nim

```diff
@@ -0,0 +1,78 @@
+## Levenshtein distance
+
+import tables
+
+type Key = (Natural, Natural)
+
+
+func toKey(indA, indB: int): Key =
+  return (indA.Natural, indB.Natural)
+
+
+func initLevenshteinDistanceMatrix(lenA, lenB: Natural): Table[Key, Natural] =
+  # Partially init the distance matrix:
+  # - Pre-fill with distances between all prefixes of `a` and an empty string
+  for indA in 0.Natural..lenA:
+    result[toKey(indA, 0)] = indA
+
+  # - Pre-fill with distances between an empty string and all prefixes of `b`
+  for indB in 1.Natural..lenB:
+    result[toKey(0, indB)] = indB
+
+
+proc fillLevenshteinDistanceMatrix(
+    distances: var Table[Key, Natural],
+    a, b: string) =
+  ## `distances[toKey(posA, posB)]` is the Levenshtein distance between
+  ## prefix of `a` with length `posA` and prefix of `b` with length `posB`
+  for indA in 0..<a.len:
+    for indB in 0..<b.len:
+      let
+        substitutionCost = (if a[indA] == b[indB]: 0 else: 1)
+        distanceIfDeleted = distances[toKey(indA, indB + 1)] + 1
+        distanceIfInserted = distances[toKey(indA + 1, indB)] + 1
```

Comment:
Deletion / Insertion is a POV
A change might be an insertion when changing string a to b, and a deletion when changing b to a.
It looks like you go from string a to b and that must be told.

Reply:
This is a minute implementation detail, same as choosing to iterate over `0..<b.len` for each of `0..<a.len` and not the other way. It's pretty obvious from the code and I wouldn't say documenting it is a "must".

Documenting the main proc as commutative is a good idea, on the other hand.

Reply:
cf. 71eaddfe5fb93a6d65628cd2811cc3e2ce47ce15.